pkg/typesystem/typesystem.go

```diff
@@ -379,6 +395,59 @@ func (t *TypeSystem) UsersetCanFastPath(relationReferences []*openfgav1.Relation
 	return true
 }
 
+// recursiveUsersetNodeCanFastpath is a helper function to determine whether the node (object#relation) has
+// - one edge back to itself
+// - other edges lead directly to terminal types (not union/intersection/exclusion).
+// - Note that user:* is considered as terminal node.
+func (t *TypeSystem) recursiveUsersetNodeCanFastpath(curAuthorizationModelNode *graph.AuthorizationModelNode, userType string) bool {
+	hasNodeBackToItself := false
+	hasCorrectTerminalType := false
+
+	// TODO: Optimize by memorizing the list of neighbourNodes
+	neighbourNodes := t.authorizationModelGraph.To(curAuthorizationModelNode.ID())
+	for neighbourNodes.Next() {
+		curNode := neighbourNodes.Node()
+		curNeighbourAuthorizationModelNode, ok := curNode.(*graph.AuthorizationModelNode)
+		if !ok {
+			// Note: this should not happen, but adding the guard nonetheless
+			return false
+		}
+		switch curNeighbourAuthorizationModelNode.NodeType() {
+		case graph.SpecificType:
+			// TODO wildcard?? should we add a new NodeType to handle this more explicitly?'
```

Comment:
https://github.com/openfga/language/pull/356 :)

Reply:
Will do it in next PR